research/fed-sm/job_configs/fedsm_prostate/app/custom/learners/supervised_monai_prostate_fedsm_learner.py

```diff
@@ -0,0 +1,392 @@
+from monai.losses import DiceLoss
+from monai.networks.nets.unet import UNet
+from networks.vgg import vgg11
+
+from nvflare.apis.dxo import DXO, DataKind, MetaKey, from_shareable
+from nvflare.apis.fl_constant import ReturnCode
+from nvflare.apis.fl_context import FLContext
+from nvflare.apis.shareable import Shareable, make_reply
+from nvflare.apis.signal import Signal
+from nvflare.app_common.app_constant import AppConstants, ValidateType
+
+
+class SupervisedMonaiProstateFedSMLearner(SupervisedMonaiProstateLearner):
+    def __init__(
+        self,
+        train_config_filename,
+        aggregation_epochs: int = 1,
+        fedsm_select_epochs: int = 1,
+        train_task_name: str = AppConstants.TASK_TRAIN,
+    ):
+        """Trainer for prostate segmentation task. It inherits from MONAI trainer.
+
+        Args:
+            train_config_filename: directory of config_3 file.
+            fedsm_select_epochs: the number of training epochs of selector model. Defaults to 1.
+            aggregation_epochs: the number of training epochs of both global and personalized models for a round. Defaults to 1.
+            train_task_name: name of the task to train the model.
+
+        Returns:
+            a Shareable with the updated local model after running `execute()`
+        """
+        SupervisedMonaiProstateLearner.__init__(
+            self,
+            train_config_filename=train_config_filename,
+            aggregation_epochs=aggregation_epochs,
+            train_task_name=train_task_name,
+        )
+        self.fedsm_person_model_epochs = aggregation_epochs
+        self.fedsm_select_model_epochs = fedsm_select_epochs
+
+    def train_config(self, fl_ctx: FLContext):
+        # Initialize superclass
+        SupervisedMonaiProstateLearner.train_config(self, fl_ctx)
+
+        engine = fl_ctx.get_engine()
+        ws = engine.get_workspace()
+        app_dir = ws.get_app_dir(fl_ctx.get_job_id())
+
+        # Initialize PTFedSMHelper
+        # personalized and selector model training epoch
+        # personalized model same as global model
+        # selector model can be different from the other two task models
+        fedsm_person_model = UNet(
+            spatial_dims=2,
+            in_channels=1,
+            out_channels=1,
+            channels=(16, 32, 64, 128, 256),
+            strides=(2, 2, 2, 2),
+            num_res_units=2,
+        ).to(self.device)
+        fedsm_select_model = vgg11(
+            num_classes=self.config_info["select_num_classes"],
+        ).to(self.device)
+        fedsm_person_criterion = DiceLoss(sigmoid=True)
+        fedsm_select_criterion = torch.nn.CrossEntropyLoss()
+        fedsm_person_optimizer = optim.Adam(fedsm_person_model.parameters(), lr=self.lr)
+        fedsm_select_optimizer = optim.Adam(
+            fedsm_select_model.parameters(), lr=self.config_info["learning_rate_select"]
+        )
+        self.fedsm_helper = SupervisedPTFedSMHelper(
```

Comment:
Should be initialized as None in in `__init__()`

Reply:
added